### PR TITLE
Change default media type to JSON

### PR DESCRIPTION
**Pull request by stuartwdouglas**
This adds a new RESTEasy provider that will produce
JSON by default. This removes the need to add
@Produces(MediaType.APPLICATION_JSON) and
@Consumes(MediaType.APPLICATION_JSON) everywhere.

The old behaviour can be retained by setting:

quarkus.resteasy-json.default-json=false

in application.properties.

### CODE REVIEW

**Review by FroMage — APPROVED** [review]
Great idea!

File: docs/src/main/asciidoc/rest-json.adoc [hunk 1]
```diff
@@ -172,9 +167,13 @@ depending on the extension you chose when initializing the project.
...
 [NOTE]
 ====
-While RESTEasy supports auto-negotiation, when using Quarkus, it is very important to define the `@Produces` and `@Consumes` annotations.
-They are analyzed at build time and Quarkus restricts the number of JAX-RS providers included in the native executable to the minimum required by the application.
-It allows to reduce the size of the native executable.
+When JSON extension is installed such as `quarkus-resteasy-jsonb` Quarkus will use this by default for most return values, unless the media type is explicitly set via
+and `@Produces` or `@Consumes` annotation (there are some exceptions for well known types, such as `String` and `File`, which default to `text/plain` and `application/octet-stream`
+respectively).
+
+If you don't want JSON my default you can set `quarkus.resteasy-json.default-json=false` and the default will change back to `application/octet-stream`. If you set this
```
Comment (Ladicek) [inline docs/src/main/asciidoc/rest-json.adoc:174]:
```suggestion
If you don't want JSON by default you can set `quarkus.resteasy-json.default-json=false` and the default will change back to `application/octet-stream`. If you set this
```

File: extensions/resteasy-jsonb/runtime/src/main/java/io/quarkus/resteasy/jsonb/QuarkusJSONBSerializer.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package io.quarkus.resteasy.jsonb;
```
[the remark this answers is not in the document]
Comment (gsmet) [reply extensions/resteasy-jsonb/runtime/src/main/java/io/quarkus/resteasy/jsonb/QuarkusJSONBSerializer.java:1]: And same for the Jackson one.

**Review by gsmet — DISMISSED** [review]
Looks like I forgot to post my review... You added the Jackson impl so that's one less thing to add but there are still a few details that needs fixing.

I can do it if you have better things to do, just ping me!

**Comment by stuartwdouglas** [comment]
Done

**Review by gastaldi — APPROVED** [review]

**Comment by gastaldi** [comment]
@stuartwdouglas apparently this PR does not support this:

```java
    @GET
    public Map<String,String> hello() {
        Map<String,String> t = new HashMap<>();
        t.put("Hey", "Ho");
        return t;
    }
```

I get the following error while hitting the endpoint: 
```
2020-10-14 00:15:35,726 ERROR [org.jbo.res.res.i18n] (executor-thread-1) RESTEASY002005: Failed executing GET /greeting: org.jboss.resteasy.core.NoMessageBodyWriterFoundFailure: Could not find MessageBodyWriter for response object of type: java.util.HashMap of media type: application/octet-stream
	at org.jboss.resteasy.core.ServerResponseWriter.lambda$writeNomapResponse$3(ServerResponseWriter.java:125)
	at org.jboss.resteasy.core.interception.jaxrs.ContainerResponseContextImpl.filter(ContainerResponseContextImpl.java:404)
	at org.jboss.resteasy.core.ServerResponseWriter.executeFilters(ServerResponseWriter.java:252)
	at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:101)
	at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:74)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeResponse(SynchronousDispatcher.java:594)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:524)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:131)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:37)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:94)
	at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)

```

**Comment by gastaldi** [comment]
I created https://github.com/quarkusio/quarkus/issues/12701

**Comment by gastaldi** [comment]
Ignore it, I forgot to add `quarkus-resteasy-jackson` to my sample project

**Pull request merged**